pytools/graph.py

```diff
@@ -199,26 +203,51 @@ class HeapEntry:
     Only needs to define :func:`pytools.graph.__lt__` according to
     <https://github.com/python/cpython/blob/8d21aa21f2cbc6d50aab3f420bb23be1d081dac4/Lib/heapq.py#L135-L138>.
     """
-    def __init__(self, node, key):
-        self.node = node
-        self.key = key
+    node: Any
+    key: Any
 
     def __lt__(self, other):
         return self.key < other.key
 
 
-def compute_topological_order(graph: Mapping[T, Iterable[T]],
-                              key: Optional[Callable[[T], Any]] = None) -> List[T]:
-    """Compute a topological order of nodes in a directed graph.
+@dataclass(frozen=True)
+class TopologicalOrderState(Generic[T]):
+    """
+    .. attribute:: scheduled_nodes
+
+        A :class:`list` of nodes that have been scheduled.
+
+    .. warning::
+
+        - Mutable updates to :attr:`scheduled_nodes`
+          results in an undefined behavior.
+    """
+    scheduled_nodes: List[T]
+
+
+def compute_topological_order_with_dynamic_key(
+        graph: Mapping[T, Iterable[T]],
+        trigger_key_update: Callable[[TopologicalOrderState[T]], bool],
+        get_key: Callable[[TopologicalOrderState[T]], Callable[[T], Any]]
+) -> List[T]:
+    """
+    Computes a topological order of nodes in a directed graph with support for
+    a dynamic keying function.
 
     :arg graph: A :class:`collections.abc.Mapping` representing a directed
         graph. The dictionary contains one key representing each node in the
         graph, and this key maps to a :class:`collections.abc.Iterable` of its
         successor nodes.
 
-    :arg key: A custom key function may be supplied to determine the order in
-        break-even cases. Expects a function of one argument that is used to
-        extract a comparison key from each node of the *graph*.
+    :arg trigger_key_update: A function called after scheduling a node in
+        *graph* that takes in an instance of :class:`TopologicalOrderState`
+        corresponding to the scheduling state at that point and returns whether
+        the comparison keys corresponding to the nodes be updated.
+
+    :arg get_key: A callable  called when *trigger_key_update*
+        returns *True*. Takes in an instance of :class:`TopologicalOrderState`
+        and returns another callable that accepts node as an argument and returns the
+        comparison key corresponding to the node.
```

Comment:
Couldn't this be a single callable that either performs the update or says "I've not done anything"?